orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
+  // nodes.
+  std::vector<Node*> nodes;
+  PipelineStageNodeGroup(const size_t stage, std::vector<Node*>& node) : stage_id(stage), nodes(std::move(node)){};
+};
+
+common::Status AddPassthroughInitializer(Graph& graph,
+                                         const NodeArg* initializer,
+                                         const std::vector<PipelineStageNodeGroup>& node_groups,
+                                         const std::vector<Node*>& send_nodes,
+                                         const std::vector<Node*>& recv_nodes) {
+  ORT_ENFORCE(node_groups.size() >= 2, "Initializer ", initializer->Name(), " is not shared across stages.");
+
+  const size_t from_stage = node_groups.front().stage_id;
+  const size_t to_stage = node_groups.back().stage_id;
+
+  ORT_ENFORCE(from_stage < to_stage, "Pass through from_stage (", from_stage,
+              ") is not less than the to_stage (", to_stage, ").");
+
+  auto dtype = initializer->TypeAsProto()->tensor_type().elem_type();
+
+  // new_node_args tracks newly created node_args in the pass through stages
+  std::vector<NodeArg*> new_node_args;
+  auto current_node_arg = const_cast<NodeArg*>(initializer);
+
+  for (auto i = from_stage; i < to_stage; ++i) {
+    // processing send node in cut i
+    auto& send_attributes = send_nodes[i]->GetMutableAttributes();
+    auto& send_element_types = send_attributes["element_types"];
+    send_element_types.add_ints(static_cast<int64_t>(dtype));
+    send_nodes[i]->MutableInputDefs().push_back(current_node_arg);
+    send_nodes[i]->MutableInputArgsCount().back()++;
+
+    // Create a new node_arg for the recv, as the new node_arg from recv node should possess a different id
+    // than the one in send
+    auto& new_node_arg = CreateNodeArg(graph, current_node_arg);
+    new_node_args.push_back(&new_node_arg);
+    current_node_arg = &new_node_arg;
+
+    // process recv node in cut i
+    auto& recv_attributes = recv_nodes[i]->GetMutableAttributes();
+    auto& recv_element_types = recv_attributes["element_types"];
+    recv_element_types.add_ints(static_cast<int64_t>(dtype));
+    recv_nodes[i]->MutableOutputDefs().push_back(current_node_arg);
+  }
+
+  // update the consumer node's input if the node's group is not in the first partition
+  for (size_t i = 1u; i < node_groups.size(); ++i) {
+    ORT_ENFORCE(node_groups[i].stage_id > from_stage, "node group id (", node_groups[i].stage_id,
+                ") is less than first stage id (", from_stage, "). ");
+    size_t new_node_arg_index = node_groups[i].stage_id - from_stage - 1;
+    for(auto node : node_groups[i].nodes){
+      for (auto& input_node : node->MutableInputDefs()) {
+        if (input_node == initializer) {
+          input_node = new_node_args[new_node_arg_index];
+          break;
+        }
+      }
+    }
+  }
+  return Status::OK();
+}
+
+void TraverseGraphWithConnectedElement(Graph& graph,
```

Comment:
I'm curious if we have any general graph traversal functions -- some of the logic here is really separate from the particular transformation.

Reply:
AFAIK, this is the only place we use an undirected graph traversal.. More commonly used function is `ReverseDFSFrom()` in graph.h, which treats the graph as directed and only checks producer node and ignore inputs and outputs. If it will be used in other place in the future, we could consider move it to a more general place. What do you think?

---
In reply to: [452275562](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452275562) [](ancestors = 452275562)